README.md
# `oauth2-mock-server`

[![npm package](https://img.shields.io/npm/v/oauth2-mock-server.svg?logo=npm)](https://www.npmjs.com/package/oauth2-mock-server)
[![Node.js version](https://img.shields.io/node/v/oauth2-mock-server.svg)](https://nodejs.org/)

> _OAuth 2 mock server. Intended to be used for development or testing purposes._

When developing an application that exposes or consumes APIs that are secured with an OAuth 2 authorization scheme, a mechanism for issuing access tokens is needed. Frequently, a developer needs to create custom code that fakes the creation of tokens for testing purposes, and these tokens cannot be properly verified, since there is no actual entity issuing those tokens.

The purpose of this package is to provide an easily configurable OAuth 2 server, that can be set up and teared down at will, and can be programatically run while performing automated tests.

> **Warning:** This tool is _not_ intended to be used as an actual OAuth 2 server. It lacks many features that would be required in a proper implementation.

## Development prerequisites

- [Node.js 12.0+](https://nodejs.org/)
- [Yarn 1.15.2+](https://classic.yarnpkg.com/lang/en/)

## How to use

### Installation

Add it to your Node.js project as a development dependency:

With yarn...

```shell
yarn add -D oauth2-mock-server
```

...or with npm

```shell
npm install --save-dev oauth2-mock-server
```

### Quickstart

Here is an example for creating and running a server instance with a single random RSA key:

```js
const { OAuth2Server } = require('oauth2-mock-server');

let server = new OAuth2Server();

// Generate a new RSA key and add it to the keystore
await server.issuer.keys.generate('RS256');

// Start the server
await server.start(8080, 'localhost');
console.log('Issuer URL:', server.issuer.url); // -> http://localhost:8080

// Do some work with the server
// ...

// Stop the server
await server.stop();
```

Any number of existing JSON-formatted keys can be added to the keystore.

```js
// Add an existing JWK key to the keystore
await server.issuer.keys.add({
  kid: 'some-key',
  alg: "RS256",
  kty: 'RSA',
  // ...
});
```

JSON Web Tokens (JWT) can be built programmatically:

```js
const request = require('request');

// Build a new token
let token = await server.issuer.buildToken();

// Call a remote API with the token
request.get(
  'https://server.example.com/api/endpoint',
  { auth: { bearer: token } },
  function callback(err, res, body) {
    /* ... */
  }
);
```

### Supported JWK formats

| Algorithm          | kty | alg                         |
| ------------------ | --- | --------------------------- |
| RSASSA-PKCS1-v1_5  | RSA | RS256, RS384, RS512         |
| RSASSA-PSS         | RSA | PS256, PS384, PS512         |
| ECDSA              |  EC | ES256, ES256K, ES384, ES512 |
| Edwards-curve DSA  | OKP | EdDSA                       |

### Customization hooks

It also provides a convenient way, through event emitters, to programmatically customize the server processing. This is particularly useful when expecting the OIDC service to behave in a specific way on one single test:

- The JWT access token

  ```js
  // Modify the expiration time on next token produced
  service.once('beforeTokenSigning', (token, _req) => {
    const timestamp = Math.floor(Date.now() / 1000);
    token.payload.exp = timestamp + 400;
  });
  ```

  ```js
  // Add the client ID to a token
  const basicAuth = require('basic-auth');
  service.once('beforeTokenSigning', (token, req) => {
    const credentials = basicAuth(req);
    const clientId = credentials ? credentials.name : req.body.client_id;
    token.payload.client_id = clientId;
  });
  ```

- The token endpoint response body and status

  ```js
  // Force the oidc service to provide an invalid_grant response on next call to the token endpoint
  service.once('beforeResponse', (tokenEndpointResponse, req) => {
    tokenEndpointResponse.body = {
      error: 'invalid_grant',
    };
    tokenEndpointResponse.statusCode = 400;
  });
  ```

- The userinfo endpoint response body and status

  ```js
  // Force the oidc service to provide an error on next call to userinfo endpoint
  service.once('beforeUserinfo', (userInfoResponse, req) => {
    userInfoResponse.body = {
      error: 'invalid_token',
      error_message: 'token is expired',
    };
    userInfoResponse.statusCode = 401;
  });
  ```

- The revoke endpoint response body and status

  ```js
  // Simulates a custom token revocation body
  service.once('beforeRevoke', (revokeResponse, req) => {
    revokeResponse.body = {
      result: 'revoked',
    };
  });
  ```

- The authorization endpoint redirect uri and query parameters

  ```js
  // Modify the uri and query parameters before the authorization redirect
  service.once('beforeAuthorizeRedirect', (authorizeRedirectUri, req) => {
    authorizeRedirectUri.url.searchParams.set('foo', 'bar');
  });
  ```

## Supported endpoints

### GET `/.well-known/openid-configuration`

Returns the [OpenID Provider Configuration Information](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig) for the server.

### GET `/jwks`

Returns the JSON Web Key Set (JWKS) of all the keys configured in the server.

### POST `/token`

Issues access tokens. Currently, this endpoint is limited to:

- No authentication
- Client Credentials grant
- Resource Owner Password Credentials grant
- Authorization code grant
- Refresh token grant

### GET `/authorize`

It simulates the user authentication. It will automatically redirect to the callback endpoint sent as parameter.
It currently supports only 'code' response_type.

### GET `/userinfo`

It provides extra userinfo claims.

### POST `/revoke`

It simulates a token revocation. This endpoint should always return 200 as stated by [RFC 7009](https://tools.ietf.org/html/rfc7009#section-2.2).

## Command-Line Interface

The server can be run from the command line. You can either install it globally:

```shell
npm install -g oauth2-mock-server
oauth2-mock-server --help
```

or run it directly:

```shell
npx oauth2-mock-server --help
```

## Attributions

- [`jose`](https://www.npmjs.com/package/jose)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [4.0.0](https://github.com/axa-group/oauth2-mock-server/compare/v3.2.0...v4.0.0) — 2021-08-TODO

### Removed

- Drop support for Node.js 10
- No longer accepts PEM encoded keys
- No longer supports generating unsigned JWTs

### Changed

- **Breaking:** Reworked exposed API. Please refer to the [migration guide](./MIGRATION.md) for more information.
- Add support for Node.js 16

## [3.2.0](https://github.com/axa-group/oauth2-mock-server/compare/v3.1.0...v3.2.0) — 2021-08-03

### Added

- Add `subject_types_supported` OpenID Provider Metadata field (thanks [jjbooth74](https://github.com/jjbooth74)!)

## [3.1.0](https://github.com/axa-group/oauth2-mock-server/compare/v3.0.3...v3.1.0) — 2020-11-30

### Added

- Add authorize redirect event (thanks [markwallsgrove](https://github.com/markwallsgrove)!)

## [3.0.3](https://github.com/axa-group/oauth2-mock-server/compare/v3.0.2...v3.0.3) — 2020-11-12

### Fixed

- Fix regression: When adding a key to the KeyStore, do not normalize key "use" value to "sig" when already defined

## [3.0.2](https://github.com/axa-group/oauth2-mock-server/compare/v3.0.1...v3.0.2) — 2020-10-29

### Added

- Support Nodejs 14.15 LTS

## [3.0.1](https://github.com/axa-group/oauth2-mock-server/compare/v3.0.0...v3.0.1) — 2020-10-23

### Fixed

- Include missing files on pack/publish

## [3.0.0](https://github.com/axa-group/oauth2-mock-server/compare/v2.0.0...v3.0.0) — 2020-10-22

### Added

- TypeScript type definitions ([#48](https://github.com/axa-group/oauth2-mock-server/pull/48))

### Changed

- Straightened definitions of optional parameters: `null` is no longer considered as a non valued parameter value; `undefined` bears that meaning.

## [2.0.0](https://github.com/axa-group/oauth2-mock-server/compare/v1.5.1...v2.0.0) — 2020-10-01

### Added

- Honor OpenID Connect `nonce` ([#34](https://github.com/axa-group/oauth2-mock-server/pull/34) by [@HASHIMOTO-Takafumi](https://github.com/HASHIMOTO-Takafumi))

### Removed

- No longer support Node 8

## [1.5.1](https://github.com/axa-group/oauth2-mock-server/compare/v1.5.0...v1.5.1) — 2020-04-06

### Security

- Update `npm` dependencies to fix:
  - [CVE-2020-7598](https://github.com/advisories/GHSA-vh95-rmgr-6w4m)

## [1.5.0](https://github.com/axa-group/oauth2-mock-server/compare/v1.4.0...v1.5.0) — 2020-01-23

### Added

- Add HTTP request object to `OAuth2Service`'s events
- Add `beforeTokenSigning` event to `OAuth2Service`

## [1.4.0](https://github.com/axa-group/oauth2-mock-server/compare/v1.3.3...v1.4.0) — 2020-01-15

### Security

- Update `npm` dependencies to fix:
  - [NPM Security Advisory 1164](https://www.npmjs.com/advisories/1164)
  - [NPM Security Advisory 1300](https://www.npmjs.com/advisories/1300)
  - [NPM Security Advisory 1316](https://www.npmjs.com/advisories/1316)
  - [NPM Security Advisory 1324](https://www.npmjs.com/advisories/1324)
  - [NPM Security Advisory 1325](https://www.npmjs.com/advisories/1325)

### Fixed

- Add missing `aud` claim under Authorization Code Flow

### Added

- Add CORS support

## [1.3.3](https://github.com/axa-group/oauth2-mock-server/compare/v1.3.2...v1.3.3) — 2019-09-25

### Security

- Update `npm` dependencies to fix:
  - [CVE-2019-15657](https://nvd.nist.gov/vuln/detail/CVE-2019-15657)
  - [CVE-2019-10746](https://nvd.nist.gov/vuln/detail/CVE-2019-10746)
  - [CVE-2019-10747](https://nvd.nist.gov/vuln/detail/CVE-2019-10747)

### Changed

- Update license's legal entity.

## [1.3.2](https://github.com/axa-group/oauth2-mock-server/compare/v1.3.1...v1.3.2) — 2019-08-09

### Security

- Update `npm` dependencies to fix:
  - [CVE-2019-10744](https://github.com/lodash/lodash/pull/4336)

## [1.3.1](https://github.com/axa-group/oauth2-mock-server/compare/v1.3.0...v1.3.1) — 2019-06-07

### Security

- Update `npm` dependencies to fix:
  - [WS-2019-0032](https://github.com/nodeca/js-yaml/issues/475)
  - [WS-2019-0063](https://github.com/nodeca/js-yaml/pull/480)
  - [WS-2019-0064](https://github.com/wycats/handlebars.js/compare/v4.1.1...v4.1.2)

## [1.3.0](https://github.com/axa-group/oauth2-mock-server/compare/v1.2.0...v1.3.0) — 2019-06-03

### Added

- Add revocation endpoint

## [1.2.0](https://github.com/axa-group/oauth2-mock-server/compare/v1.1.0...v1.2.0) — 2019-03-19

### Added

- Add Authorization code grant
- Add Refresh token grant
- Add Userinfo endpoint

### Security

- Update `npm` dependencies to fix [CVE-2018-16469](https://nvd.nist.gov/vuln/detail/CVE-2018-16469)

## [1.1.0](https://github.com/axa-group/oauth2-mock-server/compare/v1.0.0...v1.1.0) — 2018-08-02

### Added

- Add Resource Owner Password Credentials grant

### Fixed

- Add missing cache control headers on `/token` responses

## 1.0.0 — 2018-08-01

Initial release.


test/jwk-store.test.ts
import { JWKStore } from '../src/lib/jwk-store';
import * as testKeys from './keys';

describe('JWK Store', () => {
  describe('generate()', () => {
    it.each([
      ["RSASSA-PKCS1-v1_5", "RS256", "RSA"],
      ["RSASSA-PKCS1-v1_5", "RS384", "RSA"],
      ["RSASSA-PKCS1-v1_5", "RS512", "RSA"],
      ["RSASSA-PSS", "PS256", "RSA"],
      ["RSASSA-PSS", "PS384", "RSA"],
      ["RSASSA-PSS", "PS512", "RSA"],
      ["ECDSA", "ES256", "EC"],
      ["ECDSA", "ES256K", "EC"],
      ["ECDSA", "ES384", "EC"],
      ["ECDSA", "ES512", "EC"],
    ])('should be able to generate a new %s based key (alg = %s)', async (_kind: string, alg: string, expectedKty: string) => {
      const store = new JWKStore();
      const key = await store.generate(alg);
      expect(key).toMatchObject({
        alg: alg,
        kty: expectedKty,
        kid: expect.stringMatching(/^[\w-]+$/),
      });
    });

    it.each([
      "Ed25519",
      "Ed448",
    ])('should be able to generate a new EdDSA based key (crv = %s)', async (crv: string) => {
      const store = new JWKStore();
      const key = await store.generate('EdDSA', { crv });
      expect(key).toMatchObject({
        alg: 'EdDSA',
        kty: 'OKP',
        crv,
        kid: expect.stringMatching(/^[\w-]+$/),
      });
    });

    it.each([
      "RS123",
      "dunno",
    ])('throws on unsupported algs (alg = %s)', async (alg: string) => {
      const store = new JWKStore();

      await expect(() => store.generate(alg)).rejects.toThrow("unsupported or invalid JWK \"alg\" (Algorithm) Parameter value");
    });

    it.each([
      "Ed007",
      "dunno",
    ])('throws on unsupported crv for EdDSA alg (crv = %s)', async (crv: string) => {
      const store = new JWKStore();

      await expect(() => store.generate('EdDSA', { crv })).rejects.toThrow("Invalid or unsupported crv option provided, supported values are Ed25519 and Ed448");
    });

    it.each([
      ['RS256', ['e', 'n', 'd', 'p', 'q', 'dp', 'dq', 'qi']]
    ])('should return the private key of a key (alg = %s)', async (alg: string, expectedProps: string[]) => {
      const store = new JWKStore();
      const jwk = await store.generate(alg);

      for (const prop of expectedProps) {
        expect(jwk).toHaveProperty(prop);
      }
    });
  });

  describe("add()", () => {
    it.each([
      ['RSA', testKeys.getParsed('test-rs256-key.json')],
      ['EC', testKeys.getParsed('test-es256-key.json')],
      ['OKP', testKeys.getParsed('test-eddsa-key.json')],
    ])('should be able to add a JWK key to the store (kty = %s)', async (keyType, testKey) => {
      const store = new JWKStore();
      const key = await store.add(testKey);

      expect(key).toMatchObject({
        kty: keyType,
        kid: testKey.kid,
      });
    });

    it.each([
      ['RSA', testKeys.getParsed('test-rs256-key.json')],
      ['EC', testKeys.getParsed('test-es256-key.json')],
      ['OKP', testKeys.getParsed('test-eddsa-key.json')]
    ])('throws when serialized key lacks the "alg" property (kty = %s)', async (_keyType, testKey) => {
      const store = new JWKStore();

      delete testKey.alg;

      await expect(() => store.add(testKey)).rejects.toThrow('Unspecified JWK "alg" property');
    });

    it.each([
      ['RSA', testKeys.getParsed('test-rs256-key.json')],
      ['EC', testKeys.getParsed('test-es256-key.json')],
      ['OKP', testKeys.getParsed('test-eddsa-key.json')]
    ])('throws when serialized key contains an unsupported "alg" value (kty = %s)', async (_keyType, testKey) => {
      const store = new JWKStore();

      testKey.alg = "DUNNO256";

      await expect(() => store.add(testKey)).rejects.toThrow('Unsupported JWK "alg" value ("DUNNO256")');
    });

    it('throws when serialized RSA key is public"', async () => {
      const store = new JWKStore();

      const testKey = testKeys.getParsed('test-rs256-key.json');

      delete testKey.d;
      delete testKey.p;
      delete testKey.q;
      delete testKey.dp;
      delete testKey.dq;
      delete testKey.qi;

      await expect(() => store.add(testKey)).rejects.toThrow('Invalid JWK type. No "private" key related data has been found.');
    });

    it('throws when serialized EC key is public"', async () => {
      const store = new JWKStore();

      const testKey = testKeys.getParsed('test-es256-key.json');

      delete testKey.d;

      await expect(() => store.add(testKey)).rejects.toThrow('Invalid JWK type. No "private" key related data has been found.');
    });


    it('throws when serialized OKP key is public"', async () => {
      const store = new JWKStore();

      const testKey = testKeys.getParsed('test-eddsa-key.json');

      delete testKey.d;

      await expect(() => store.add(testKey)).rejects.toThrow('Invalid JWK type. No "private" key related data has been found.');
    });

    it('adding a key will overwrite an existing key in the store bearing the same "kid"', async () => {
      const store = new JWKStore();

      const one = testKeys.getParsed('test-rs256-key.json');
      expect(one.kty).toEqual("RSA");
      one.kid = "new_id";
      await store.add(one);

      const retrievedOne = store.get("new_id");
      expect(retrievedOne).not.toBeNull();
      expect(retrievedOne!.kty).toEqual(one.kty);

      const two = testKeys.getParsed('test-es256-key.json');
      expect(two.kty).toEqual("EC");
      two.kid = "new_id";
      await store.add(two);

      const retrievedTwo = store.get("new_id");
      expect(retrievedTwo).not.toBeNull();
      expect(retrievedTwo!.kty).toEqual(two.kty);
    });
  });

  describe("get()", () => {
    it('should be able to retrieve a key by its \'kid\'', async () => {
      const store = new JWKStore();
      const key1 = await store.generate('RS256', { kid: 'key-one' });
      const key2 = await store.generate('RS256', { kid: 'key-two' });

      expect(key1.kid).not.toEqual(key2.kid);

      const stored1 = store.get('key-one');
      const stored2a = store.get('key-two');
      const stored2b = store.get('key-two');
      const stored3 = store.get('non-existing-kid');

      expect(stored1).toBe(key1);
      expect(stored2a).toBe(key2);
      expect(stored2b).toBe(key2);
      expect(stored3).toBeUndefined();
    });

    it('should be able to retrieve keys in a round-robin manner', async () => {
      const store = new JWKStore();
      await store.generate('RS256', { kid: 'key-one' });
      await store.generate('RS256', { kid: 'key-two' });
      await store.generate('RS256', { kid: 'key-three' });

      const key1 = store.get();
      expect(key1).not.toBeNull();

      const key2 = store.get();
      expect(key2).not.toBeNull();
      expect(key2!.kid).not.toEqual(key1!.kid);

      const key3 = store.get();
      expect(key3).not.toBeNull();
      expect(key3!.kid).not.toEqual(key1!.kid);
      expect(key3!.kid).not.toEqual(key2!.kid);

      const key4 = store.get();
      expect(key4).not.toBeNull();
      expect(key4!.kid).toEqual(key1!.kid);
    });

    it('should return undefined when trying to retrieve a key from an empty store', () => {
      const store = new JWKStore();

      const res1 = store.get();
      const res2 = store.get('non-existing-kid');

      expect(res1).toBeUndefined();
      expect(res2).toBeUndefined();
    });
  });

  describe("toJSON()", () => {
    it.each([
      undefined,
      true,
      false
    ])('should be able to produce a JSON representation of the public keys in the key store (including private fields: %s)', async (shouldIncludePrivates?: boolean) => {
      const store = new JWKStore();
      await store.generate('RS256', { kid: 'key-one' });
      await store.generate('RS256', { kid: 'key-two' });
      await store.generate('RS256', { kid: 'key-three' });

      const jwks = store.toJSON(shouldIncludePrivates);
      expect(jwks).toHaveProperty("keys");
      expect(jwks).toBeInstanceOf(Array);

      expect(jwks).toHaveLength(3);

      for (const key of jwks) {
        expect(key).toBeInstanceOf(Object);
        expect(key).toHaveProperty("kid");
        expect(typeof (key as Record<string, unknown>).kid).toBe("string");
      }

      expect(jwks.map((key) => key.kid).sort()).toEqual(['key-one', 'key-three', 'key-two']);

      jwks.forEach((jwk) => {
        expect(store.get(jwk.kid)).not.toBeNull();

        ['e', 'n'].forEach((prop) => {
          expect(jwk).toHaveProperty(prop);
        });

        ['d', 'p', 'q', 'dp', 'dq', 'qi'].forEach((prop) => {
          const isExposed = prop in jwk;
          const shouldBeExposed = shouldIncludePrivates === true;
          expect(isExposed).toEqual(shouldBeExposed);
        });
      });
    });

    it.each([
      ['RSA', testKeys.getParsed('test-rs256-key.json'), ['d', 'p', 'q', 'dp', 'dq', 'qi']],
      ['EC', testKeys.getParsed('test-es256-key.json'), ['d']],
      ['OKP', testKeys.getParsed('test-eddsa-key.json'), ['d']]
    ])('properly removes private fields from key when requesting it (kty = %s)', async (_keyType, testKey, fields) => {
      const store = new JWKStore();

      await store.add(testKey);
      const keys = store.toJSON(false);
      expect(keys).toHaveLength(1);

      const key = keys[0];

      fields.forEach(prop => {
        expect(key).not.toHaveProperty(prop);
      });
    });
  });
});


test/lib/test_helpers.ts
import { AssertionError } from "assert";
import { JWTVerifyResult } from "jose/types";
import { parseJwk } from "jose/jwk/parse";
import { jwtVerify } from "jose/jwt/verify";

import { OAuth2Issuer } from "../../src/lib/oauth2-issuer";

export const verifyTokenWithKey = async (issuer: OAuth2Issuer, token: string, kid: string): Promise<JWTVerifyResult> => {
  const key = issuer.keys.get(kid);

  if (key === undefined) {
    throw new AssertionError({ message: 'Key is undefined' });
  }

  const privateKey = await parseJwk(key);

  const verified = await jwtVerify(token, privateKey);
  return verified;
};


src/lib/types.ts
import { JWK } from 'jose/types';
export type { JWK } from 'jose/types';

export interface TokenRequest {
  scope?: string;
  grant_type: string;
  username?: unknown;
  client_id?: unknown;
  code?: string;
}

export interface Options {
  host?: string;
  port: number;
  keys: JWK[];
  saveJWK: boolean;
}

export interface MutableAuthorizeRedirectUri {
  url: URL;
}

export interface MutableToken {
  header: Header;
  payload: Payload;
}

export interface Header {
  kid: string;
  [key: string]: unknown;
}

export interface Payload {
  iss: string;
  iat: number;
  exp: number;
  nbf: number;
  [key: string]: unknown;
}

export interface StatusCodeMutableResponse {
  statusCode: number;
}

export interface MutableResponse extends StatusCodeMutableResponse {
  body: Record<string, unknown> | '';
}

export type ScopesOrTransform = string | string[] | JwtTransform;

export interface JwtTransform {
  (header: Header, payload: Payload): void;
}

export enum PublicEvents {
  BeforeTokenSigning = 'beforeTokenSigning',
  BeforeResponse = 'beforeResponse',
  BeforeUserinfo = 'beforeUserinfo',
  BeforeRevoke = 'beforeRevoke',
  BeforeAuthorizeRedirect = 'beforeAuthorizeRedirect',
}

export enum InternalEvents {
  BeforeSigning = 'beforeSigning',
}

export interface TokenBuildOptions {
  /**
   * The 'kid' of the key that will be used to sign the JWT.
   * If omitted, the next key in the round - robin will be used.
   */
  kid?: string;

  /**
   * A scope, array of scopes, or JWT transformation callback.
   */
  scopesOrTransform?: ScopesOrTransform;

  /**
   * Time in seconds before the JWT to expire. Default: 3600 seconds.
   */
  expiresIn?: number;
}


src/lib/oauth2-service.ts
/**
 * Copyright (c) AXA Assistance France
 *
 * Licensed under the AXA Assistance France License (the "License"); you
 * may not use this file except in compliance with the License.
 * A copy of the License can be found in the LICENSE.md file distributed
 * together with this file.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * OAuth2 Service library
 *
 * @module lib/oauth2-service
 */

import { IncomingMessage } from 'http';
import express, { RequestHandler, Express } from 'express';
import cors from 'cors';
import basicAuth from 'basic-auth';
import { EventEmitter } from 'events';
import { v4 as uuidv4 } from 'uuid';

import { OAuth2Issuer } from './oauth2-issuer';
import { assertIsString, assertIsValidTokenRequest } from './helpers';
import type {
  JwtTransform,
  MutableAuthorizeRedirectUri,
  MutableResponse,
  MutableToken,
  ScopesOrTransform,
  StatusCodeMutableResponse,
} from './types';
import { InternalEvents, PublicEvents } from './types';

const OPENID_CONFIGURATION_PATH = '/.well-known/openid-configuration';
const TOKEN_ENDPOINT_PATH = '/token';
const JWKS_URI_PATH = '/jwks';
const AUTHORIZE_PATH = '/authorize';
const USERINFO_PATH = '/userinfo';
const REVOKE_PATH = '/revoke';

/**
 * Provides a request handler for an OAuth 2 server.
 */
export class OAuth2Service extends EventEmitter {
  /**
   * Creates a new instance of OAuth2Server.
   *
   * @param {OAuth2Issuer} oauth2Issuer The OAuth2Issuer instance
   *     that will be offered through the service.
   */

  #issuer: OAuth2Issuer;
  #requestHandler: Express;
  #nonce: Record<string, string>;

  constructor(oauth2Issuer: OAuth2Issuer) {
    super();
    this.#issuer = oauth2Issuer;

    this.#requestHandler = this.buildRequestHandler();

    this.#nonce = {};
  }

  /**
   * Returns the OAuth2Issuer instance bound to this service.
   *
   * @type {OAuth2Issuer}
   */
  get issuer(): OAuth2Issuer {
    return this.#issuer;
  }

  /**
   * Builds a JWT with a key in the keystore. The key will be selected in a round-robin fashion.
   *
   * @param {IncomingMessage} req The incoming HTTP request.
   * @param {number} expiresIn Time in seconds for the JWT to expire. Default: 3600 seconds.
   * @param {ScopesOrTransform} [scopesOrTransform] A scope, array of scopes,
   *     or JWT transformation callback.
   * @returns {Promise<string>} The produced JWT.
   * @fires OAuth2Service#beforeTokenSigning
   */
  async buildToken(
    req: IncomingMessage,
    expiresIn: number,
    scopesOrTransform: ScopesOrTransform | undefined
  ): Promise<string> {
    this.issuer.once(InternalEvents.BeforeSigning, (token: MutableToken) => {
      /**
       * Before token signing event.
       *
       * @event OAuth2Service#beforeTokenSigning
       * @param {MutableToken} token The unsigned JWT header and payload.
       * @param {IncomingMessage} req The incoming HTTP request.
       */
      this.emit(PublicEvents.BeforeTokenSigning, token, req);
    });

    return await this.issuer.buildToken({ scopesOrTransform, expiresIn });
  }

  /**
   * Returns a request handler to be used as a callback for http.createServer().
   *
   * @type {Function}
   */
  get requestHandler(): Express {
    return this.#requestHandler;
  }

  private buildRequestHandler = () => {
    const app = express();
    app.disable('x-powered-by');
    app.use(cors());
    app.get(OPENID_CONFIGURATION_PATH, this.openidConfigurationHandler);
    app.get(JWKS_URI_PATH, this.jwksHandler);
    app.post(
      TOKEN_ENDPOINT_PATH,
      express.urlencoded({ extended: false }),
      this.tokenHandler
    );
    app.get(AUTHORIZE_PATH, this.authorizeHandler);
    app.get(USERINFO_PATH, this.userInfoHandler);
    app.post(REVOKE_PATH, this.revokeHandler);

    return app;
  };

  private openidConfigurationHandler: RequestHandler = (_req, res) => {
    assertIsString(this.issuer.url, 'Unknown issuer url.');

    const openidConfig = {
      issuer: this.issuer.url,
      token_endpoint: `${this.issuer.url}${TOKEN_ENDPOINT_PATH}`,
      authorization_endpoint: `${this.issuer.url}${AUTHORIZE_PATH}`,
      userinfo_endpoint: `${this.issuer.url}${USERINFO_PATH}`,
      token_endpoint_auth_methods_supported: ['none'],
      jwks_uri: `${this.issuer.url}${JWKS_URI_PATH}`,
      response_types_supported: ['code'],
      grant_types_supported: [
        'client_credentials',
        'authorization_code',
        'password',
      ],
      token_endpoint_auth_signing_alg_values_supported: ['RS256'],
      response_modes_supported: ['query'],
      id_token_signing_alg_values_supported: ['RS256'],
      revocation_endpoint: `${this.issuer.url}${REVOKE_PATH}`,
      subject_types_supported: ['public'],
    };

    return res.json(openidConfig);
  };

  private jwksHandler: RequestHandler = (_req, res) => {
    res.json({ keys: this.issuer.keys.toJSON() });
  };

  private tokenHandler: RequestHandler = async (req, res) => {
    const tokenTtl = 3600;

    res.set({
      'Cache-Control': 'no-store',
      Pragma: 'no-cache',
    });

    let xfn: ScopesOrTransform | undefined;

    assertIsValidTokenRequest(req.body);
    const reqBody = req.body;

    let { scope } = reqBody;

    switch (req.body.grant_type) {
      case 'client_credentials':
        xfn = scope;
        break;
      case 'password':
        xfn = (_header, payload) => {
          Object.assign(payload, {
            sub: reqBody.username,
            amr: ['pwd'],
            scope,
          });
        };
        break;
      case 'authorization_code':
        scope = 'dummy';
        xfn = (_header, payload) => {
          Object.assign(payload, {
            sub: 'johndoe',
            amr: ['pwd'],
            scope,
          });
        };
        break;
      case 'refresh_token':
        scope = 'dummy';
        xfn = (_header, payload) => {
          Object.assign(payload, {
            sub: 'johndoe',
            amr: ['pwd'],
            scope,
          });
        };
        break;
      default:
        return res.status(400).json({
          error: 'invalid_grant',
        });
    }

    const token = await this.buildToken(req, tokenTtl, xfn);
    const body: Record<string, unknown> = {
      access_token: token,
      token_type: 'Bearer',
      expires_in: tokenTtl,
      scope,
    };
    if (req.body.grant_type !== 'client_credentials') {
      const credentials = basicAuth(req);
      const clientId = credentials ? credentials.name : req.body.client_id;

      const xfn: JwtTransform = (_header, payload) => {
        Object.assign(payload, {
          sub: 'johndoe',
          aud: clientId,
        });
        if (reqBody.code !== undefined && this.#nonce[reqBody.code]) {
          Object.assign(payload, {
            nonce: this.#nonce[reqBody.code],
          });
          delete this.#nonce[reqBody.code];
        }
      };

      body.id_token = await this.buildToken(req, tokenTtl, xfn);
      body.refresh_token = uuidv4();
    }

    const tokenEndpointResponse: MutableResponse = {
      body,
      statusCode: 200,
    };

    /**
     * Before token response event.
     *
     * @event OAuth2Service#beforeResponse
     * @param {MutableResponse} response The response body and status code.
     * @param {IncomingMessage} req The incoming HTTP request.
     */
    this.emit(PublicEvents.BeforeResponse, tokenEndpointResponse, req);

    return res
      .status(tokenEndpointResponse.statusCode)
      .json(tokenEndpointResponse.body);
  };

  private authorizeHandler: RequestHandler = (req, res) => {
    const { scope, state } = req.query;
    const responseType = req.query.response_type;
    const redirectUri = req.query.redirect_uri;
    const code = uuidv4();

    let queryNonce: string | undefined;

    if ('nonce' in req.query) {
      assertIsString(req.query.nonce, 'Invalid nonce type');
      queryNonce = req.query.nonce;
    }

    assertIsString(redirectUri, 'Invalid redirectUri type');
    assertIsString(scope, 'Invalid scope type');
    assertIsString(state, 'Invalid state type');

    const url = new URL(redirectUri);

    if (responseType === 'code') {
      if (queryNonce !== undefined) {
        this.#nonce[code] = queryNonce;
      }
      url.searchParams.set('code', code);
      url.searchParams.set('scope', scope);
      url.searchParams.set('state', state);
    } else {
      url.searchParams.set('error', 'unsupported_response_type');
      url.searchParams.set(
        'error_description',
        'The authorization server does not support obtaining an access token using this response_type.'
      );
      url.searchParams.set('state', state);
    }

    const authorizeRedirectUri: MutableAuthorizeRedirectUri = { url };

    /**
     * Before authorize redirect event.
     *
     * @event OAuth2Service#beforeAuthorizeRedirect
     * @param {MutableAuthorizeRedirectUri} authorizeRedirectUri The redirect uri and query params to redirect to.
     * @param {IncomingMessage} req The incoming HTTP request.
     */
    this.emit(PublicEvents.BeforeAuthorizeRedirect, authorizeRedirectUri, req);

    // Note: This is a textbook definition of an "open redirect" vuln
    // cf. https://cwe.mitre.org/data/definitions/601.html
    //
    // However, this whole library is expected to be used as a test helper,
    // so there's no real point in making the exposed API more complex (by
    // exposing an endpoint to preregister whitelisted urls, for instance)
    // for the sake of security.
    //
    // This is *not* a real oAuth2 server. This is *not* to be run in production.
    res.redirect(url.href); // lgtm[js/server-side-unvalidated-url-redirection]
  };

  private userInfoHandler: RequestHandler = (req, res) => {
    const userInfoResponse: MutableResponse = {
      body: {
        sub: 'johndoe',
      },
      statusCode: 200,
    };

    /**
     * Before user info event.
     *
     * @event OAuth2Service#beforeUserinfo
     * @param {MutableResponse} response The response body and status code.
     * @param {IncomingMessage} req The incoming HTTP request.
     */
    this.emit(PublicEvents.BeforeUserinfo, userInfoResponse, req);

    res.status(userInfoResponse.statusCode).json(userInfoResponse.body);
  };

  private revokeHandler: RequestHandler = (req, res) => {
    const revokeResponse: StatusCodeMutableResponse = {
      statusCode: 200,
    };

    /**
     * Before revoke event.
     *
     * @event OAuth2Service#beforeRevoke
     * @param {StatusCodeMutableResponse} response The response status code.
     * @param {IncomingMessage} req The incoming HTTP request.
     */
    this.emit(PublicEvents.BeforeRevoke, revokeResponse, req);

    return res.status(revokeResponse.statusCode).send('');
  };
}
